### PR TITLE
Fuzzy search for tools search

### PR DESCRIPTION
**Pull request by anuprulez**
This PR proposes to include approximate string search (also called fuzzy search) for tools. It is implemented using the same query divided into 3-grams. For example a string 'text' has these 3-grams - 'tex' and 'ext'. The q-grams size is kept 3 as the number of minimum characters required to perform the search is also 3.

With each 3-gram, the search is performed as in the existing approach and the score for each result (tool) given by BM25 is added up for each 3-gram. The results are then sorted in the decreasing order of their BM25 scores. The main advantage of using this approximate search comes when a query is misspelt. For example, if one wants to type 'text' as query but misspell it and type 'texx'. In this scenario, the proposed approach still shows the following tools:

> 'addValue', 'mergeCols1', 'genomespace_importer', 'comp1', 'Convert characters1', 'trimmer', 'createInterval', 'Remove beginning1', 'cat1', 'Show beginning1', 'Cut1', 'ChangeCase', 'secure_hash_message_digest', 'wc_gnu', 'Show tail1', 'Paste1', 'random_lines1', 'Grep1', 'Count1', 'gff_filter_by_attribute'

while the existing approach brings none.

With the correct spelling of 'text', these results are obtained:

proposed approach - 

> 'addValue', 'mergeCols1', 'gene2exon1', 'Convert characters1', 'Show beginning1', 'Cut1', 'wc_gnu', 'random_lines1', 'ChangeCase', 'secure_hash_message_digest', 'Paste1', 'trimmer', 'createInterval', 'cat1', 'Show tail1', 'Extract_features1', 'Extract genomic DNA 1', 'Interval2Maf1', 'genomespace_importer', 'Interval2Maf_pairwise1', 'maf_by_block_number1', 'comp1', 'Remove beginning1', 'Grep1', 'Count1', 'gff_filter_by_attribute'

existing approach -

> 'addValue', 'mergeCols1', 'genomespace_importer', 'comp1', 'wc_gnu', 'random_lines1', 'ChangeCase', 'trimmer', 'createInterval', 'Show tail1', 'secure_hash_message_digest', 'Convert characters1', 'Cut1', 'Paste1', 'cat1', 'Show beginning1', 'Remove beginning1', 'Grep1', 'Count1', 'gff_filter_by_attribute'

Another example for query 'charcter':

> The 3-grams are 'cha', 'har', 'arc', 'rct', 'cte', 'ter'

proposed approach - 

> 'trimmer', 'Grep1', 'wc_gnu', 'barchart_gnuplot', 'Filter1', 'gff_filter_by_attribute', 'ucsc_table_direct_archaea1', 'sort1', 'ChangeCase', 'MAF_To_Interval1', 'gtf_filter_by_attribute_values_list', 'gff_filter_by_feature_count', 'Convert characters1', 'Extract_features1', 'upload1', 'Interval2Maf_pairwise1', 'MAF_filter', 'gtf2bedgraph', 'Interval_Maf_Merged_Fasta2', 'wiggle2simple1', 'Extract genomic DNA 1', 'Summary_Statistics1', 'createInterval', 'Interval2Maf1', 'genomespace_exporter', 'wig_to_bigWig', 'MAF_To_Fasta1', 'join1', 'MAF_To_BED1', 'Count1', 'secure_hash_message_digest', 'Paste1', 'gff2bed1'

existing approach - 

> No result

The q-gram size can be made configurable. Your suggestions are welcome.

### CODE REVIEW

**Comment by nsoranzo** [comment]
Nice! That's been requested multiple times!

There shouldn't be any change in static/ since you didn't modify any javascript or css file.

File: lib/galaxy/tools/search/__init__.py [hunk 1]
```diff
@@ -105,9 +105,40 @@ def search( self, q, tool_name_boost, tool_section_boost, tool_description_boost
...
+                            # Add the current score with the previous one if the
+                            # tool appears again for the next qgram
+                            hits_with_score[ prev_hit ] = curr_hits.score(i) + hits_with_score[ prev_hit ]
+                    # Add the tool if not present to the collection with its score
+                    if not is_present:
+                        hits_with_score[ curr_hit['id'] ] = curr_hits.score(i)
+            # Sort the results based on aggregated BM25 score in decreasing order of scores
+            hits_with_score = sorted(hits_with_score.items(), key=lambda x: x[1], reverse=True)
+            # Return the tool ids
+            return [item[0] for item in hits_with_score]
```
Comment (martenson) [inline lib/galaxy/tools/search/__init__.py]: This does not respect the `tool_search_limit` configuration. It returns all hits.

File: lib/galaxy/tools/search/__init__.py [hunk 2]
```diff
@@ -105,9 +105,40 @@ def search( self, q, tool_name_boost, tool_section_boost, tool_description_boost
         # Hyphens are wildcards in Whoosh causing bad things
         if q.find( '-' ) != -1:
             q = (' ').join( [ token.text for token in self.rex( to_unicode( q ) ) ] )
-        # Perform the search
-        hits = searcher.search( parser.parse( '*' + q + '*' ), limit=float( tool_search_limit ) )
-        return [ hit[ 'id' ] for hit in hits ]
+        # Fuzzy search using q-grams
+        qgram_length = 3
```
Comment (martenson) [inline lib/galaxy/tools/search/__init__.py]: Making `qgram_length` configurable is a good idea. I would pursue it.

**Review by martenson — CHANGES_REQUESTED** [review]
Thank you for the PR. Apart from the inline ones I have 2 main comments:

1. Given this considerably changes output of the API and the behavior of tool search I would prefer this to be a configurable option, at least until we are sure it is superior to fulltext.

2. For ngrams searching we could use the Whoosh built-in ngram support (http://whoosh.readthedocs.io/en/latest/ngrams.html) which includes `NgramFilter` for ngram splitting with min and maxsize.

Thank you for the contribution, I hope my review is helpful to you.

**Comment by martenson** [comment]
@galaxybot test this

**Comment by anuprulez** [comment]
@martenson Thanks a lot for your review!

The changes include:

- `NgramFilter` from Whoosh search library for breaking down the query into multiple ngrams.

- Ngrams search is now configurable through a config variable `tool_enable_ngram_search` in `galaxy.ini` file. Moreover, the minimum and maximum sizes of ngrams can be configured as well through `tool_ngram_minsize` and `tool_ngram_maxsize` variables respectively.

- For limiting the search results upto the value of `tool_ngram_minsize` in Ngrams search, the items are taken from starting to this value in the set of hits.

@bgruening

**Comment by martenson** [comment]
@galaxybot test this

**Review by martenson — APPROVED** [review]

**Comment by martenson** [comment]
Good job @anuprulez - looking great! Thank you for the contribution very much.

**Comment by anuprulez** [comment]
Thanks a lot for your suggestions and merging!
@martenson @bgruening @nsoranzo

**Comment by bgruening** [comment]
Wuhu!! Thanks to all! Good job @anuprulez!

**Pull request merged**